src/sinks/util/http.rs

```diff
@@ -17,20 +21,92 @@ use tower_hyper::client::Client;
 use tracing::field;
 use tracing_tower::{InstrumentableService, InstrumentedService};
 
-pub type RequestBuilder = Box<dyn Fn(Vec<u8>) -> hyper::Request<Vec<u8>> + Sync + Send>;
 pub type Response = hyper::Response<Bytes>;
 pub type Error = hyper::Error;
 
+pub trait HttpSink: Send + Sync + 'static {
+    type Input;
+    type Output;
+
+    fn encode_event(&self, event: Event) -> Option<Self::Input>;
+    fn build_request(&self, events: Self::Output) -> http::Request<Vec<u8>>;
+}
+
+pub struct BatchedHttpSink<T, B: Batch>
+where
+    B::Output: Clone,
+{
+    sink: Arc<T>,
+    inner: BatchSink<B, TowerBatchedSink<B::Output, HttpRetryLogic, B, HttpService<B::Output>>>,
+    // An empty slot is needed to buffer an item where we encoded it but
+    // the inner sink is applying back pressure. This trick is used in the `WithFlatMap`
+    // sink combinator. https://docs.rs/futures/0.1.29/src/futures/sink/with_flat_map.rs.html#20
+    slot: Option<B::Input>,
+}
+
+impl<T, B> BatchedHttpSink<T, B>
+where
+    B: Batch,
+    B::Output: Clone,
+    T: HttpSink<Input = B::Input, Output = B::Output>,
+{
+    pub fn new(
+        sink: T,
+        batch: B,
+        request_settings: TowerRequestSettings,
+        batch_settings: BatchSettings,
+        tls_settings: Option<TlsSettings>,
+        cx: &SinkContext,
+    ) -> Self {
+        let sink = Arc::new(sink);
+        let sink1 = sink.clone();
+        let svc = HttpService::new2(cx.resolver(), tls_settings, move |b| sink1.build_request(b));
+
+        let service_sink = request_settings.batch_sink(HttpRetryLogic, svc, cx.acker());
+        let inner = BatchSink::from_settings(service_sink, batch, batch_settings);
```

Comment:
This is rather difficult to follow (I know not all these methods are new, just pointing it out). A quick read seems to imply we're creating a batch sink on the first line and then wrapping it with another batch sink on the second line?

Reply:
Yeah, we need to nest `BatchSink<BatchServiceSink<Service, Batch>>>` but since our service stack is pretty large I moved that to a type alias.